### PR TITLE
Attempt to fix some concurrency issues

### PR DESCRIPTION
**Pull request by mliszcz**
Possible fix for #511

### CODE REVIEW

**Comment by gscalamera** [comment]
Thanks @mliszcz for the patch, it seems to fix the #511 issue but I have some concerns about the use of atomics:

1) I am not sure they are (at least not all of them) needed, for example notifd_event and zmq_event in commit a45325355cc3fb34b2bc9f79b0cf38b722161c6a should be used sequentially in the code (written when subscribing and read when firing events), so no data race should be possible even without synchronization
2) Avoiding atomics would keep this code still C++11 free to be easily ported in 9.3-backports branch

BTW I've been running the test in #511 for a couple of days both with this patch and without the commits introducing the atomics with no segmentation till now.

**Comment by mliszcz** [comment]
Hi @gscalamera,
Thanks for the re-test and good news!

As for the atomics - I have not investigated much but I also think they are not necessary. Writing `bool` should be atomic on x86 anyway without special synchronization on the application level.  I've changed the flags to atomic just to silence tsan warnings (neither of the threads is holding any lock):

```
==================
WARNING: ThreadSanitizer: data race (pid=858)
  Read of size 1 at 0x7b7000002eb7 by thread T8:
    #0 Tango::Attribute::use_zmq_event() /home/michal/Documents/cppTango/cppapi/server/attribute.h:2234 (libtango.so.9+0x7a8e2c)
    #1 Tango::Attribute::fire_change_event(Tango::DevFailed*) /home/michal/Documents/cppTango/cppapi/server/attribute.cpp:3794 (libtango.so.9+0x79ffc1)
    #2 Tango::DeviceImpl::push_change_event(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, double*, long, long, bool) /home/michal/Documents/cppTango/cppapi/server/dev_event.cpp:842 (libtango.so.9+0x7d993f)
    #3 TestPushCrash_ns::readthread::run(void*) ../src/readthread.cpp:131 (TestPushCrash+0x1e0e4)
    #4 omni_thread_wrapper /home/michal/Documents/omniORB-4.2.3/src/lib/omnithread/posix.cc:459 (libomnithread.so.4+0x3f34)
    #5 <null> <null> (libtsan.so.0+0x296ad)

  Previous write of size 1 at 0x7b7000002eb7 by thread T13:
    #0 Tango::Attribute::set_use_zmq_event() /home/michal/Documents/cppTango/cppapi/server/attribute.h:2263 (libtango.so.9+0x8a01f1)
    #1 Tango::DServer::event_subscription(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, Tango::ChannelType, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, int&, int&, Tango::DeviceImpl*, int) /home/michal/Documents/cppTango/cppapi/server/eventcmds.cpp:468 (libtango.so.9+0x89bca2)
    #2 Tango::DServer::zmq_event_subscription_change(Tango::DevVarStringArray const*) /home/michal/Documents/cppTango/cppapi/server/eventcmds.cpp:922 (libtango.so.9+0x89e0bf)
    #3 Tango::ZmqEventSubscriptionChangeCmd::execute(Tango::DeviceImpl*, CORBA::Any const&) /home/michal/Documents/cppTango/cppapi/server/dserverclass.cpp:1370 (libtango.so.9+0x87d879)
```

I also do not think that protecting store_in_bb or event_cptr is important in this scenario.

Could you please do one more test run and see if it is enough to patch these two files?:
* cppapi/server/eventsupplier.cpp
* cppapi/server/attribute.cpp

I need to re-do the PR anyway and I'd like to focus on the most probable root-cause.

**Comment by gscalamera** [comment]
Tested during the weekend with only eventsupplier.cpp and attribute.cpp patched and no segmentation till now.

File: cppapi/server/attribute.cpp [hunk 1]
```diff
@@ -4024,12 +4029,16 @@ void Attribute::fire_change_event(DevFailed *except)
...
 			std::vector<double> filterable_data;
 			std::vector<std::string> filterable_names_lg;
 			std::vector<long> filterable_data_lg;
 
+			{
+			omni_mutex_lock oml(EventSupplier::get_event_mutex());
+
 			if (except != NULL)
 			{
 				prev_change_event.err    = true;
```
Comment (mliszcz) [inline cppapi/server/attribute.cpp]: * protecting access to prev_change_event, prev_archive_event and prev_quality_event
* again, we are taking the lock from another class to protect own data, it this ok?
* maybe we should add `Attribute::set_prev_change_event(const LastAttrValue&)` which will lock some mutex? Which mutex?

Race between readthread, here in this scope (line 4054 in this PR, no mutex) and polling thread (eventsupplier.cpp:416) - polling thread holds attr_mutex and event_mutex:
https://github.com/tango-controls/cppTango/blob/6a2857c32baba5454d936815279314e60b838b9b/cppapi/server/attribute.cpp#L4043-L4046
https://github.com/tango-controls/cppTango/blob/6a2857c32baba5454d936815279314e60b838b9b/cppapi/server/eventsupplier.cpp#L412-L414

```
==================
WARNING: ThreadSanitizer: data race (pid=3222)
  Write of size 8 at 0x7b040000a1a0 by thread T8:
    #0 operator delete[](void*) <null> (libtsan.so.0+0x742ad)
    #1 _CORBA_Sequence<double>::freebuf(double*) /usr/include/omniORB4/seqTemplatedecls.h:89 (libtango.so.9+0x5258fd)
    #2 _CORBA_Sequence<double>::~_CORBA_Sequence() /usr/include/omniORB4/seqTemplatedecls.h:127 (libtango.so.9+0x5281d6)
    #3 _CORBA_Unbounded_Sequence<double>::~_CORBA_Unbounded_Sequence() /usr/include/omniORB4/seqTemplatedecls.h:253 (libtango.so.9+0x524c2b)
    #4 _CORBA_Unbounded_Sequence_w_FixSizeElement<double, 8, 8>::~_CORBA_Unbounded_Sequence_w_FixSizeElement() /usr/include/omniORB4/seqTemplatedecls.h:516 (libtango.so.9+0x52196b)
    #5 Tango::DevVarDoubleArray::~DevVarDoubleArray() /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:233 (libtango.so.9+0x51b625)
    #6 Tango::AttrValUnion::_release_member() /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:3281 (libtango.so.9+0x592cf5)
    #7 Tango::AttrValUnion::double_att_value(Tango::DevVarDoubleArray const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:3036 (libtango.so.9+0x592100)
    #8 Tango::AttrValUnion::operator=(Tango::AttrValUnion const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:2860 (libtango.so.9+0x591947)
    #9 Tango::Attribute::fire_change_event(Tango::DevFailed*) /home/michal/Documents/cppTango/cppapi/server/attribute.cpp:4054 (libtango.so.9+0x79f312)
    #10 Tango::DeviceImpl::push_change_event(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, double*, timeval&, Tango::AttrQuality, long, long, bool) /home/michal/Documents/cppTango/cppapi/server/dev_event.cpp:1120 (libtango.so.9+0x7d903c)
    #11 TestPushCrash_ns::readthread::run(void*) ../src/readthread.cpp:147 (TestPushCrash+0x34403)
    #12 omni_thread_wrapper /home/michal/Documents/omniORB-4.2.3/src/lib/omnithread/posix.cc:459 (libomnithread.so.4+0x3f34)
    #13 <null> <null> (libtsan.so.0+0x296ad)

  Previous write of size 8 at 0x7b040000a1a0 by thread T9 (mutexes: write M1262, write M1004):
    #0 operator new[](unsigned long) <null> (libtsan.so.0+0x73fda)
    #1 _CORBA_Sequence<double>::allocbuf(unsigned int) /usr/include/omniORB4/seqTemplatedecls.h:84 (libtango.so.9+0x528964)
    #2 _CORBA_Sequence<double>::copybuffer(unsigned int) /usr/include/omniORB4/seqTemplatedecls.h:205 (libtango.so.9+0x525767)
    #3 _CORBA_Sequence<double>::length(unsigned int) /usr/include/omniORB4/seqTemplatedecls.h:59 (libtango.so.9+0x52298a)
    #4 _CORBA_Sequence<double>::_CORBA_Sequence(_CORBA_Sequence<double> const&) /usr/include/omniORB4/seqTemplatedecls.h:147 (libtango.so.9+0x5ac977)
    #5 _CORBA_Unbounded_Sequence<double>::_CORBA_Unbounded_Sequence(_CORBA_Unbounded_Sequence<double> const&) /usr/include/omniORB4/seqTemplatedecls.h:242 (libtango.so.9+0x5a4867)
    #6 _CORBA_Unbounded_Sequence_w_FixSizeElement<double, 8, 8>::_CORBA_Unbounded_Sequence_w_FixSizeElement(_CORBA_Unbounded_Sequence_w_FixSizeElement<double, 8, 8> const&) /usr/include/omniORB4/seqTemplatedecls.h:503 (libtango.so.9+0x599ceb)
    #7 Tango::DevVarDoubleArray::DevVarDoubleArray(Tango::DevVarDoubleArray const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:238 (libtango.so.9+0x590c43)
    #8 Tango::AttrValUnion::double_att_value(Tango::DevVarDoubleArray const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:3040 (libtango.so.9+0x592162)
    #9 Tango::AttrValUnion::operator=(Tango::AttrValUnion const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:2860 (libtango.so.9+0x591947)
    #10 Tango::EventSupplier::detect_and_push_change_event(Tango::DeviceImpl*, Tango::EventSupplier::SuppliedEventData&, Tango::Attribute&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, Tango::DevFailed*, bool) /home/michal/Documents/cp
pTango/cppapi/server/eventsupplier.cpp:416 (libtango.so.9+0x89dfa4)
```

File: cppapi/server/eventsupplier.cpp [hunk 2]
```diff
@@ -766,11 +770,13 @@ bool EventSupplier::detect_and_push_archive_event(DeviceImpl *device_impl,
...
+            omni_mutex_lock oml(detect_mutex);
             attr.prev_archive_event.err = true;
             attr.prev_archive_event.except = *except;
         }
         else
         {
+            omni_mutex_lock oml(detect_mutex);
             if (attr_value.attr_val_5 != NULL)
             {
                 attr.prev_archive_event.value_4 = attr_value.attr_val_5->value;
```
Comment (mliszcz) [inline cppapi/server/eventsupplier.cpp]: * probably should be applied also to prev_change_event

Race is between polling thread (line 782 in PR, holds attr_mutex and event_mutex) and readthread (line 1837 in PR, holds detect_mutex):
https://github.com/tango-controls/cppTango/blob/6a2857c32baba5454d936815279314e60b838b9b/cppapi/server/eventsupplier.cpp#L1830

```
==================                                                                                                                                                                                                                                   (1/29 results) [1399/14345]
WARNING: ThreadSanitizer: data race (pid=3222)
  Read of size 8 at 0x7b040000a320 by thread T8 (mutexes: write M1006):
    #0 Tango::EventSupplier::detect_change(Tango::Attribute&, Tango::EventSupplier::SuppliedEventData&, bool, double&, double&, Tango::DevFailed*, bool&, Tango::DeviceImpl*) /home/michal/Documents/cppTango/cppapi/server/eventsupplier.cpp:1837 (libtango.so.9+0x8a4898)
    #1 Tango::Attribute::fire_archive_event(Tango::DevFailed*) /home/michal/Documents/cppTango/cppapi/server/attribute.cpp:4484 (libtango.so.9+0x7a0c8c)
    #2 Tango::DeviceImpl::push_archive_event(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, double*, timeval&, Tango::AttrQuality, long, long, bool) /home/michal/Documents/cppTango/cppapi/server/dev_event.cpp:1764 (libtango.so.9+0x7dafda)
    #3 TestPushCrash_ns::readthread::run(void*) ../src/readthread.cpp:148 (TestPushCrash+0x34480)
    #4 omni_thread_wrapper /home/michal/Documents/omniORB-4.2.3/src/lib/omnithread/posix.cc:459 (libomnithread.so.4+0x3f34)
    #5 <null> <null> (libtsan.so.0+0x296ad)

  Previous write of size 8 at 0x7b040000a320 by thread T9 (mutexes: write M1262, write M1004):
    #0 operator new[](unsigned long) <null> (libtsan.so.0+0x73fda)
    #1 _CORBA_Sequence<double>::allocbuf(unsigned int) /usr/include/omniORB4/seqTemplatedecls.h:84 (libtango.so.9+0x528964)
    #2 _CORBA_Sequence<double>::copybuffer(unsigned int) /usr/include/omniORB4/seqTemplatedecls.h:205 (libtango.so.9+0x525767)
    #3 _CORBA_Sequence<double>::length(unsigned int) /usr/include/omniORB4/seqTemplatedecls.h:59 (libtango.so.9+0x52298a)
    #4 _CORBA_Sequence<double>::_CORBA_Sequence(_CORBA_Sequence<double> const&) /usr/include/omniORB4/seqTemplatedecls.h:147 (libtango.so.9+0x5ac977)
    #5 _CORBA_Unbounded_Sequence<double>::_CORBA_Unbounded_Sequence(_CORBA_Unbounded_Sequence<double> const&) /usr/include/omniORB4/seqTemplatedecls.h:242 (libtango.so.9+0x5a4867)
    #6 _CORBA_Unbounded_Sequence_w_FixSizeElement<double, 8, 8>::_CORBA_Unbounded_Sequence_w_FixSizeElement(_CORBA_Unbounded_Sequence_w_FixSizeElement<double, 8, 8> const&) /usr/include/omniORB4/seqTemplatedecls.h:503 (libtango.so.9+0x599ceb)
    #7 Tango::DevVarDoubleArray::DevVarDoubleArray(Tango::DevVarDoubleArray const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:238 (libtango.so.9+0x590c43)
    #8 Tango::AttrValUnion::double_att_value(Tango::DevVarDoubleArray const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:3040 (libtango.so.9+0x592162)
    #9 Tango::AttrValUnion::operator=(Tango::AttrValUnion const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:2860 (libtango.so.9+0x591947)
    #10 Tango::EventSupplier::detect_and_push_archive_event(Tango::DeviceImpl*, Tango::EventSupplier::SuppliedEventData&, Tango::Attribute&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, Tango::DevFailed*, timeval*, bool) /home/michal/D
ocuments/cppTango/cppapi/server/eventsupplier.cpp:782 (libtango.so.9+0x89f797)
    #11 Tango::EventSupplier::detect_and_push_events(Tango::DeviceImpl*, Tango::EventSupplier::SuppliedEventData&, Tango::DevFailed*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, timeval*) /home/michal/Documents/cppTango/cppapi/server/
eventsupplier.cpp:269 (libtango.so.9+0x89d4f7)
    #12 Tango::PollThread::poll_attr(Tango::WorkItem&) /home/michal/Documents/cppTango/cppapi/server/pollthread.cpp:1968 (libtango.so.9+0x90bd84)
    #13 Tango::PollThread::one_more_poll() /home/michal/Documents/cppTango/cppapi/server/pollthread.cpp:799 (libtango.so.9+0x9061ee)
    #14 Tango::PollThread::run_undetached(void*) /home/michal/Documents/cppTango/cppapi/server/pollthread.cpp:192 (libtango.so.9+0x903276)
    #15 omni_thread_wrapper /home/michal/Documents/omniORB-4.2.3/src/lib/omnithread/posix.cc:462 (libomnithread.so.4+0x3f6a)
    #16 <null> <null> (libtsan.so.0+0x296ad)
```

Race is also between polling thread (line 782 in PR, holds attr_mutex and event_mutex) and readthread here (no mutex):
https://github.com/tango-controls/cppTango/blob/6a2857c32baba5454d936815279314e60b838b9b/cppapi/server/attribute.cpp#L4490-L4492

```
==================
WARNING: ThreadSanitizer: data race (pid=3222)
  Read of size 1 at 0x7b0800013628 by thread T9 (mutexes: write M1262, write M1004):
    #0 _CORBA_Sequence<double>::~_CORBA_Sequence() /usr/include/omniORB4/seqTemplatedecls.h:127 (libtango.so.9+0x52818d)
    #1 _CORBA_Unbounded_Sequence<double>::~_CORBA_Unbounded_Sequence() /usr/include/omniORB4/seqTemplatedecls.h:253 (libtango.so.9+0x524c2b)
    #2 _CORBA_Unbounded_Sequence_w_FixSizeElement<double, 8, 8>::~_CORBA_Unbounded_Sequence_w_FixSizeElement() /usr/include/omniORB4/seqTemplatedecls.h:516 (libtango.so.9+0x52196b)
    #3 Tango::DevVarDoubleArray::~DevVarDoubleArray() /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:233 (libtango.so.9+0x51b625)
    #4 Tango::AttrValUnion::_release_member() /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:3281 (libtango.so.9+0x592cf5)
    #5 Tango::AttrValUnion::double_att_value(Tango::DevVarDoubleArray const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:3036 (libtango.so.9+0x592100)
    #6 Tango::AttrValUnion::operator=(Tango::AttrValUnion const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:2860 (libtango.so.9+0x591947)
    #7 Tango::EventSupplier::detect_and_push_archive_event(Tango::DeviceImpl*, Tango::EventSupplier::SuppliedEventData&, Tango::Attribute&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, Tango::DevFailed*, timeval*, bool) /home/michal/Do
cuments/cppTango/cppapi/server/eventsupplier.cpp:782 (libtango.so.9+0x89f797)
    #8 Tango::EventSupplier::detect_and_push_events(Tango::DeviceImpl*, Tango::EventSupplier::SuppliedEventData&, Tango::DevFailed*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, timeval*) /home/michal/Documents/cppTango/cppapi/server/e
ventsupplier.cpp:269 (libtango.so.9+0x89d4f7)
    #9 Tango::PollThread::poll_attr(Tango::WorkItem&) /home/michal/Documents/cppTango/cppapi/server/pollthread.cpp:1968 (libtango.so.9+0x90bd84)
    #10 Tango::PollThread::one_more_poll() /home/michal/Documents/cppTango/cppapi/server/pollthread.cpp:799 (libtango.so.9+0x9061ee)
    #11 Tango::PollThread::run_undetached(void*) /home/michal/Documents/cppTango/cppapi/server/pollthread.cpp:192 (libtango.so.9+0x903276)
    #12 omni_thread_wrapper /home/michal/Documents/omniORB-4.2.3/src/lib/omnithread/posix.cc:462 (libomnithread.so.4+0x3f6a)
    #13 <null> <null> (libtsan.so.0+0x296ad)

  Previous write of size 8 at 0x7b0800013628 by thread T8:
    #0 operator new(unsigned long) <null> (libtsan.so.0+0x73f0a)
    #1 Tango::AttrValUnion::double_att_value(Tango::DevVarDoubleArray const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:3040 (libtango.so.9+0x592150)
    #2 Tango::AttrValUnion::operator=(Tango::AttrValUnion const&) /home/michal/Documents/cppTango/cppapi/server/idl/tango.h:2860 (libtango.so.9+0x591947)
    #3 Tango::Attribute::fire_archive_event(Tango::DevFailed*) /home/michal/Documents/cppTango/cppapi/server/attribute.cpp:4505 (libtango.so.9+0x7a0d4b)
    #4 Tango::DeviceImpl::push_archive_event(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, double*, timeval&, Tango::AttrQuality, long, long, bool) /home/michal/Documents/cppTango/cppapi/server/dev_event.cpp:1764 (libtango.so.9+0x7dafda)
    #5 TestPushCrash_ns::readthread::run(void*) ../src/readthread.cpp:148 (TestPushCrash+0x34480)
    #6 omni_thread_wrapper /home/michal/Documents/omniORB-4.2.3/src/lib/omnithread/posix.cc:459 (libomnithread.so.4+0x3f34)
    #7 <null> <null> (libtsan.so.0+0x296ad)
```

**Comment by Ingvord** [comment]
> We may go the easy way and merge what we already have, but I don't think placing the locks around in random places will do any good in a long term.

+1

The thing is though,  from my experience, any clever refactoring may end up in enormous efforts. So if some one is ready to spend considerable amount of time, of course it would be much better to extract and encapsulate as much as possible

**Comment by bourtemb** [comment]
I think the main issue we have right now is that it is not documented what data a specific mutex is supposed to protect.
The name of the mutex variable may help in some situations. I would guess the following:
attr_mutex seems to protect some data related to attributes.
EventSupplier::get_event_mutex() seems to protect data related to events to be sent.
detect_mutex seems to protect data related to attribute alarm detection.

File: cppapi/server/attribute.cpp [hunk 3]
```diff
@@ -3773,9 +3773,12 @@ void Attribute::fire_change_event(DevFailed *except)
 		time_t change3_subscription,change4_subscription,change5_subscription;
 
 		now = time(NULL);
+		{
+		omni_mutex_lock oml(EventSupplier::get_event_mutex());
 		change3_subscription = now - event_change3_subscription;
 		change4_subscription = now - event_change4_subscription;
 		change5_subscription = now - event_change5_subscription;
+        }
```
[the remark this answers is not in the document]
Comment (bourtemb) [reply cppapi/server/attribute.cpp]: >     * do we patch everything in place (like in current PR) or should we create some methods on attribute which will do the locking:

I am in favour of adding new methods in this situation.

**Comment by mliszcz** [comment]
Thanks @Ingvord  and @bourtemb 

> attr_mutex seems to protect some data related to attributes.
> EventSupplier::get_event_mutex() seems to protect data related to events to be sent.
> detect_mutex seems to protect data related to attribute alarm detection.
@bourtemb, from the implementation:
* ```cpp
  omni_mutex attr_mutex; // Mutex to protect the attributes shared data buffer
  ```
* `detect_mutex`'s sole purpose is to protect whole `EventSupplier::detect_change`
* `event_mutex` - is usually used to protect various data in Attribute / Device, including event_xxx_subscription and prev_xxx_event. I do not know if this is intended use case for this mutex.

So, to correct races from my previous comment, we should use `event_mutex`. Unfortunetly, using event_mutex in `EventSupplier::detect_change` causes "lock inversion" problem (possible deadlock):
https://github.com/tango-controls/cppTango/blob/6a2857c32baba5454d936815279314e60b838b9b/cppapi/server/eventsupplier.cpp#L1830

I'll try to come with a solution.

**Comment by mliszcz** [comment]
@bourtemb, @Ingvord, I've spent some time evaluating possible options to resolve this PR. To start with something simple (which can be backported to 9.3-backports if needed), I propose the following:
* All code that sets previous event value (like in below example) in attribute.cpp and eventsupplier.cpp, will be protected by `event_mutex`. This could be extracted to a method in attribute to avoid duplications, but this method must not lock the mutex (otherwise we will lock it twice in some execution paths)
  ```cpp
    if (attr.prev_archive_event.inited == false)
    {
        if (except != NULL)
        {
            attr.prev_archive_event.err = true;
            attr.prev_archive_event.except = *except;
        }
        else
        {
            if (attr_value.attr_val_5 != NULL)
            {
                attr.prev_archive_event.value_4 = attr_value.attr_val_5->value;
            }
   ```
* We will remove `detect_mutex`. Currently it protects whole `EventSupplier::detect_change`. The whole `detect_change` should be protected by the `event_mutex` - all this function does is basically reading and comparing previous event value (and that value should be protected by `event_mutex` - see previous point and previous post). No lock will be acquired inside detect_change. `detect_change` is called from three places: `EventSupplier::detect_and_push_change_event`, `EventSupplier::detect_and_push_archive_event` and `Attribute::fire_archive_event`. First two already hold `event_mutex` when they call `detect_change`. I'll just add lock to the last one, like below:
   ```cpp
            {
			omni_mutex_lock oml(EventSupplier::get_event_mutex());
            if (event_supplier_nd != NULL)
                event_supplier_nd->detect_change(*this, ad,true,delta_change_rel,delta_change_abs,except,force_change,dev);
            else if (event_supplier_zmq != NULL)
                event_supplier_zmq->detect_change(*this, ad,true,delta_change_rel,delta_change_abs,except,force_change,dev);
            }
   ```
* Setting timestamps is unlikely to cause any trouble, but I'll wrap all access in event_mutex as well:
   ```cpp
		{
		omni_mutex_lock oml(EventSupplier::get_event_mutex());
		archive3_subscription = now - event_archive3_subscription;
		archive4_subscription = now - event_archive4_subscription;
		archive5_subscription = now - event_archive5_subscription;
		}
   ```

A potential way towards better encapsulation of `Attribute::prev_xxx_event` (which is a pre-requisite for more robust locking) is to start with cleaning `EventSupplier::detect_change`, namely: remove duplicated code (checking for all data types one by one) and remove `GET_SEQ` macro (which is used to read attribute value from the previous event). But this could be addressed in a separate issue / PR.

Are you ok with proposed changes?

**Comment by mliszcz** [comment]
The other PR is merged, let's close this one.

**Pull request closed without merging**